.github/pull_request_template.md

```diff
@@ -0,0 +1,53 @@
+Fixes #XXXX.
+
+(Replace XXXX with the id of the issue fixed in this PR. Remove this line if there is no corresponding
+issue. Don't reference the issue in the title of this pull-request.)
+
+Add tags to your PR if you are a committer (only committers have the right to add tags). Add [Design Review] tag
+if this PR should better be reviewed by at least two people.
+Don't forget to add the following tags (if applicable): [Incompatible], [Release Notes], [Compatibility], [Security],
+[Development Blocker]. Add at least one [Area - ] tag, consider creating a new one if none of the existing [Area - ]
+tags is applicable.
+
+### Description
+
+Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's
+not necessary to repeat the description here, however, you may choose to keep one summary sentence.
+
+Describe your patch: what did you change in code? How did you fix the problem?
+
+If there are several relatively logically separate changes in this PR, list them. For example:
+ - Fixed the bug ...
+ - Renamed the class ...
+ - Added a forbidden-apis entry ...
+
+Some of the aspects mentioned above may be omitted for simple and small PRs.
+
+### Design
+
+Please describe any design decisions made, including:
+ - Choice of algorithms
+ - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such
+   as when insufficient resources are available?
+ - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
+ - Method organization and design (how the logic is split between methods, parameters and return types)
+ - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
+
+In addition, describe _at least one_ alternative design (or mention alternative name) for every design (or naming)
+decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen).
+
+If you already did this in the associated issue (e. g. a "Proposal" issue), leave the following sentence:
+
+Design of this change is discussed [here](<link to Github issue or comment where you discuss the design>).
+
+This section may be omitted for really simple and small patches. However, any patch that adds a new production class
+almost certainly shouldn't omit this section.
+
+<hr>
+
+I've self-reviewed this PR (including using the [concurrency checklist](
+https://github.com/leventov/java-concurrency-checklist)).
+
+Leave the sentence above if you've self-reviewed your PR. Leave the part in parens if your PR has any relation to Java
```

Comment:
I wonder if instead of instructing people to leave a sentence, we should incorporate a markdown checklist we recommend people fill out, with high level things things to do, including self review, maybe something like:

This PR has been:
- [ ] Self-reviewed ([recommended review checklist](
https://github.com/leventov/java-concurrency-checklist)).
- [ ] Added documentation for new or modified features or behaviors
- [ ] Added Javadocs to public and non-trivial members
- [ ] Added code comments for hard to understand areas
- [ ] Added unit tests or modified existing tests to cover new code paths
- [ ] Added integration tests
- [ ] Tested in a test environment
- [ ] Tested in a production environment

I don't know that all these things are necessary, or if anything is missing, and I think we would need to make clear that not all of these things are required. I was thinking mostly that it's a way for an author to convey to reviewers what has gone into the PR so far. We could suggest that PR authors are free to remove checks which are not applicable as well. Or maybe this is too much? I'm not sure, just throwing out ideas.

Reply:
I really like this idea.

Reply:
I like the spirit of a list like this but I think we need to be careful with the wording to avoid being confusing to newish contributors. It seems to suggest that the list of things is required for every PR. In fact, it's more like a list of things to consider doing if appropriate, rather than a list of requirements. Not every change needs to be tested in a production environment. Not every change needs integration tests, and in fact it's probably counter-productive to try (for lots of changes, the slowdown from adding integration tests isn't worth it).

Reply:
I definitely agree that we would need to make clear that some items are optional and probably not recommended for most PRs or may not always be applicable, but as a reviewer this idea has grown on me quite a bit and I think I am totally for it in some form since I think it might help review go more smoothly. I would think the first 5 things are highly recommended, and 2-4 are probably required, integration tests if the behavior being or modified is very complex and has too many parts to capture in unit tests, and testing in clusters totally optional as a way for the author to communicate their own confidence in that the PR is in a good place and works as intended, should the reviewers approve the implementation.

Reply:
Hmm, ok. This looks better to me. Maybe it would be better to add a non-applicable section for items which are not applicable and so ppl can move those items to that section?